domain/relation/state/relation.go

```diff
@@ -1631,6 +1643,23 @@ WHERE  name = $getUnit.name
 	}
 
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// If the unit is already in scope, return nil to be idempotent.
+		// Settings will not be changed. Even when the unit or relation is
+		// dying, EnterScope must succeed if already entered. This allows
+		// for the relation tear down hook flow to happen.
+		var count rows
+		exists := nameAndUUID{
+			Name: unitName.String(),
+			UUID: relationUUID.String(),
+		}
+		err = tx.Query(ctx, alreadyInScopeStmt, exists).Get(&count)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Capture(err)
+		}
+		if count.Count == 1 {
+			return relationerrors.RelationUnitAlreadyExists
+		}
```

Comment:
I'd probably be slightly more defensive here and state that it's not equal to zero.

Reply:
Can only happen if someone makes a bad change to the SQL for the table by removing:

```
CREATE UNIQUE INDEX idx_relation_unit
ON relation_unit (relation_endpoint_uuid, unit_uuid);
```

Allowing for more than one is not correct either.